### PR TITLE
fix: ignore hidden elements for keyboard navigation

### LINKED ISSUE #1938: bug: it's possible to focus invisible elements

### PR DESCRIPTION
**Pull request by jeripeierSBB**
Closes #1938

### CODE REVIEW

File: src/global/a11y/interactivity-checker.spec.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    it('should return false for the child of a `visibility: hidden` element', () => {
+      testContainerElement.innerHTML = `<div style="visibility: hidden;">
+           <input>
+         </div>`;
+      const input = testContainerElement.querySelector('input') as HTMLElement;
+
+      expect(checker.isVisible(input)).toBe(false);
+    });
+
+    it.skip(
```
Comment (jeripeierSBB) [inline src/global/a11y/interactivity-checker.spec.ts:51]: maybe write as e2e test somehow.

File: src/global/a11y/focus.ts [hunk 2]
```diff
@@ -38,6 +40,15 @@ export function getFocusableElements(
   return [...focusableEls];
 }
 
+// TODO: optimize (performance) finding first element.
```
Comment (DavideMininni-Fincons) [inline src/global/a11y/focus.ts:43]: What about this TODO? does it needs more work, or it can be removed?

File: src/components/sbb-navigation/sbb-navigation.tsx [hunk 3]
```diff
@@ -328,6 +328,7 @@ export class SbbNavigation implements ComponentInterface {
           'sbb-navigation-section[data-state="opening"], sbb-navigation-section[data-state="opened"]',
         );
         if (!isBreakpoint('zero', 'large')) {
+          // TODO: Check if it works with invisible elements
```
Comment (DavideMininni-Fincons) [inline src/components/sbb-navigation/sbb-navigation.tsx:331]: What about this TODO? does it needs more work, or it can be removed?

File: src/components/sbb-breadcrumb-group/sbb-breadcrumb-group.tsx [hunk 4]
```diff
@@ -135,6 +135,7 @@ export class SbbBreadcrumbGroup implements ComponentInterface {
     evt: KeyboardEvent,
     breadcrumbs: HTMLSbbBreadcrumbElement[] = this._breadcrumbs,
   ): void {
+    // TODO: focus only visible elements
```
Comment (DavideMininni-Fincons) [inline src/components/sbb-breadcrumb-group/sbb-breadcrumb-group.tsx:138]: What about this TODO? does it needs more work, or it can be removed?

File: src/components/sbb-tab-group/sbb-tab-group.tsx [hunk 5]
```diff
@@ -101,14 +101,16 @@ export class SbbTabGroup implements ComponentInterface {
...
+      (t) => !isValidAttribute(t, 'disabled') && interactivityChecker.isVisible(t),
+    );
   }
 
   public connectedCallback(): void {
     this._isNested = !!hostContext('sbb-tab-group', this._element);
   }
 
-  public componentWillLoad(): void {
+  public componentDidLoad(): void {
```
Comment (DavideMininni-Fincons) [inline src/components/sbb-tab-group/sbb-tab-group.tsx:113]: Why this change is needed?

File: src/global/a11y/interactivity-checker.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+class InteractivityCheckerMock implements InterfaceInteractivityChecker {
+  public isVisible(): boolean {
+    return true;
+  }
+}
+
+export const interactivityChecker: InterfaceInteractivityChecker =
+  getComputedStyle(document.documentElement).visibility === ''
+    ? new InteractivityCheckerMock()
+    : new InteractivityChecker();
```
Comment (DavideMininni-Fincons) [inline src/global/a11y/interactivity-checker.ts:38]: No tests on this?

**Review by DavideMininni-Fincons — COMMENTED** [review]
Good work 👍 some comments added

File: src/components/sbb-tab-group/sbb-tab-group.tsx (hunk 5 above)
Comment (dauriamarco) [reply src/components/sbb-tab-group/sbb-tab-group.tsx:113]: To be able to use the `getComputedStyle` and make the interactivity check :)

File: src/global/a11y/interactivity-checker.ts (hunk 6 above)
Comment (dauriamarco) [reply src/global/a11y/interactivity-checker.ts:38]: It's hard to have a proper element in the test environment (with an actual visibility and width/height) and mocking all the values doesn't really make sense.

File: src/components/sbb-tab-group/sbb-tab-group.e2e.ts [hunk 7]
```diff
@@ -83,5 +83,45 @@ describe('sbb-tab-group', () => {
...
+  });
+
+  it('activates the first enabled tab if exceeds the length of the tab group', async () => {
+    page = await newE2EPage();
+    await page.setContent(`<sbb-tab-group initial-selected-index="2">
+              <sbb-tab-title id="sbb-tab-1">Test tab label 1</sbb-tab-title>
+              <div>Test tab content 1</div>
+              <sbb-tab-title id="sbb-tab-2">Test tab label 2</sbb-tab-title>
+              <div>Test tab content 2</div>
+            </sbb-tab-group>`);
```
Comment (DavideMininni-Fincons) [inline src/components/sbb-tab-group/sbb-tab-group.e2e.ts:104]: Could be better formatted, like
```suggestion
    await page.setContent(`
        <sbb-tab-group initial-selected-index="2">
          <sbb-tab-title id="sbb-tab-1">Test tab label 1</sbb-tab-title>
          <div>Test tab content 1</div>
          <sbb-tab-title id="sbb-tab-2">Test tab label 2</sbb-tab-title>
          <div>Test tab content 2</div>
        </sbb-tab-group>
    `);
```
Same for next test.

**Review by DavideMininni-Fincons — CHANGES_REQUESTED** [review]
Just a couple details, but very small, then it could be merged :)

**Review by dauriamarco — APPROVED** [review]
LGTM 👍

File: src/global/a11y/interactivity-checker.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+   * being clipped by an `overflow: hidden` parent or being outside the viewport.
+   *
+   * @returns Whether the element is visible.
+   */
+  public isVisible(element: HTMLElement): boolean {
+    return hasGeometry(element) && getComputedStyle(element).visibility === 'visible';
+  }
+}
+
+class InteractivityCheckerMock implements InterfaceInteractivityChecker {
```
Comment (dauriamarco) [inline src/global/a11y/interactivity-checker.ts:29]: I was surprised to see that the check works for the e2e tests, which is good. At this point I wonder if the Mock function is still necessary. Could it be removed?

Comment (jeripeierSBB) [reply src/global/a11y/interactivity-checker.ts:29]: The mock is needed for the unit tests.

**Review by DavideMininni-Fincons — APPROVED** [review]
LGTM, thanks!

File: src/components/sbb-tab-group/sbb-tab-group.tsx [hunk 9]
```diff
@@ -283,7 +285,7 @@ export class SbbTabGroup implements ComponentInterface {
   }
 
   @Listen('keydown')
-  public handleKeyDown(evt: KeyboardEvent): void {
+  public async handleKeyDown(evt: KeyboardEvent): Promise<void> {
```
Comment (kyubisation) [inline src/components/sbb-tab-group/sbb-tab-group.tsx:288]: What is the purpose of this change?

**Review by kyubisation — APPROVED** [review]
LGTM 👍

File: src/components/sbb-tab-group/sbb-tab-group.tsx (hunk 9 above)
Comment (jeripeierSBB) [reply src/components/sbb-tab-group/sbb-tab-group.tsx]: Don't know, removed it :)
